src/providers/eth-estimate-gas-provider.ts

```diff
@@ -0,0 +1,127 @@
+import { BigNumber } from '@ethersproject/bignumber';
+import { JsonRpcProvider } from '@ethersproject/providers';
+
+import { SwapOptions, SwapRoute, SwapType } from '../routers';
+import { ChainId, log } from '../util';
+import {
+  calculateGasUsed,
+  initSwapRouteFromExisting,
+} from '../util/gas-factory-helpers';
+
+import { ProviderConfig } from './provider';
+import { SimulationStatus, Simulator } from './simulation-provider';
+import { IV2PoolProvider } from './v2/pool-provider';
+import { ArbitrumGasData, OptimismGasData } from './v3/gas-data-provider';
+import { IV3PoolProvider } from './v3/pool-provider';
+
+export class EthEstimateGasSimulator extends Simulator {
+  v2PoolProvider: IV2PoolProvider;
+  v3PoolProvider: IV3PoolProvider;
+  constructor(
+    chainId: ChainId,
+    provider: JsonRpcProvider,
+    v2PoolProvider: any,
+    v3PoolProvider: any
+  ) {
+    super(provider, chainId);
+    this.v2PoolProvider = v2PoolProvider;
+    this.v3PoolProvider = v3PoolProvider;
+  }
+  private async ethEstimateGas(
+    fromAddress: string,
+    swapOptions: SwapOptions,
+    route: SwapRoute,
+    l2GasData?: ArbitrumGasData | OptimismGasData
+  ): Promise<SwapRoute> {
+    const currencyIn = route.trade.inputAmount.currency;
+    let estimatedGasUsed: BigNumber;
+    if (swapOptions.type == SwapType.UNIVERSAL_ROUTER) {
+      log.info(
+        { methodParameters: route.methodParameters },
+        'Simulating using eth_estimateGas on Universal Router'
+      );
+      estimatedGasUsed = await this.provider.estimateGas({
+        data: route.methodParameters!.calldata,
+        to: route.methodParameters!.to,
+        from: fromAddress,
+        value: BigNumber.from(
+          currencyIn.isNative ? route.methodParameters!.value : '0'
+        ),
+      });
+    } else if (swapOptions.type == SwapType.SWAP_ROUTER_02) {
+      log.info(
+        { methodParameters: route.methodParameters },
+        'Simulating using eth_estimateGas on SwapRouter02'
+      );
+
+      estimatedGasUsed = await this.provider.estimateGas({
+        data: route.methodParameters!.calldata,
+        to: route.methodParameters!.to,
+        from: fromAddress,
+        value: BigNumber.from(
+          currencyIn.isNative ? route.methodParameters!.value : '0'
+        ),
+      });
```

Comment:
Shouldn't try/catch this and return `simulationStatus: SimulationStatus.Failed` here if this errors out? i.e. We attempted to do a simulation and something went wrong

Reply:
okay, i'll add try/catch. I think I remember in the past we decided to not try/catch because we expect provider.estimateGas to work. I think it makes sense to try/catch now.